### PR TITLE
crystal: init at 0.20.3

### PR DESCRIPTION
**Pull request by sifmelcara**
###### Motivation for this change
There is a package request requests the Crystal programming language #21247 .

This pr adds two packages:
crystal: the Crystal compiler
shards: dependency manager for Crystal language

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

All sample code provided by crystal is tested.
All sample code works except sdl related sample.
(~~I'm not sure if this is caused by my packaging~~
ok. It appears that it also do not work on archlinux, at least for me)
Edit: sdl issue for reference: crystal-lang/crystal#3678 

~~I will test these on other linux distro and i686 when I have time.~~ Done.
Tested on archlinux x86_64 and ubuntu 16.04 i386.

cc @mimadrid

All suggestions are welcomed.

### CODE REVIEW

File: pkgs/development/compilers/crystal/default.nix [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      crystal-${version}-1/embedded/bin/crystal
+    patchelf --set-rpath ${ stdenv.lib.makeLibraryPath [ stdenv.cc.cc ] } \
+      crystal-${version}-1/embedded/bin/crystal
+  '';
+
+  buildPhase = ''
+    ${fixPrebuiltBinary}
+
+    cd crystal-${version}
+    make PATH="../crystal-${version}-1/bin:$PATH"
```
Comment (mimadrid) [inline pkgs/development/compilers/crystal/default.nix]: You are the expert here but perhaps it is better build the compiler in release mode with release=1. 
https://github.com/crystal-lang/crystal/blob/master/Makefile

**Review by LnL7 — APPROVED** [review]

**Review by mimame — APPROVED** [review]

**Comment by sifmelcara** [comment]
travis-ci build fails because the compilation takes too long and no output from crystal compiler.
I wonder if hydra will also kill this? If it does, I will try to find some method to make crystal compiler output something while compiling.

**Comment by mimadrid** [comment]
@sifmelcara, the 0.20.3 version was released several minutes ago with _optimized compile times, specially on linux_ perhaps this help a bit. https://github.com/crystal-lang/crystal/releases/tag/0.20.3

Best regards.

File: pkgs/development/compilers/crystal/default.nix [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  libPath = stdenv.lib.makeLibraryPath ([
+    boehmgc libatomic_ops pcre libevent
+  ] ++ stdenv.lib.optionals stdenv.isDarwin [
+    libiconv
+  ]);
+
+  unpackPhase = ''
+    tar zxf ${src}
+    tar zxf ${prebuiltBinary}
```
Comment (Mic92) [inline pkgs/development/compilers/crystal/default.nix]: So native compilation is not possible here?

Comment (Mic92) [inline pkgs/development/compilers/crystal/default.nix]: ok. I understand they bootstrap the compiler by using a previous version of the compiler.

**Comment by sifmelcara** [comment]
updated to the latest version of crystal (0.20.3).
PR title also updated.

**Comment by sifmelcara** [comment]
After update crystal 0.20.3, the compilation still takes a half hour on my notebook (without output anything).
If it is acceptable to compile this package on hydra without output to terminal, I think this pr is ready to be merged.
Thanks for review!

**Comment by LnL7** [comment]
I'm not sure, but I've seen hydra timing out builds that don't produce any output for some time (something like 7200s).
But we can merge this and fix it afterwards if it turns out hydra doesn't like it.

**Review by LnL7 — APPROVED** [review]

**Pull request merged**